### PR TITLE
[AArch64] Guard against getRegisterBitWidth returning zero in vector instr cost.

### PR DESCRIPTION
**Pull request by davemgreen**
If the getRegisterBitWidth is zero (such as in sme streaming functions), then we could hit a crash from using % RegWidth.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 1]
```diff
@@ -3248,19 +3248,18 @@ InstructionCost AArch64TTIImpl::getVectorInstrCostHelper(
...
+      const auto *BO = dyn_cast<BinaryOperator>(EEUser);
       return BO && BO->getOpcode() == BinaryOperator::FMul &&
              !BO->getType()->isVectorTy();
     };
 
     // Check if the extract index is from lane 0 or lane equivalent to 0 for a
     // certain scalar type and a certain vector register width.
-    auto IsExtractLaneEquivalentToZero = [&](const unsigned &Idx,
-                                             const unsigned &EltSz) {
+    auto IsExtractLaneEquivalentToZero = [&](unsigned Idx, unsigned EltSz) {
```
Comment (sdesmalen-arm) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3258]: Could you land the NFC changes separately?

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 2]
```diff
@@ -3277,13 +3276,15 @@ InstructionCost AArch64TTIImpl::getVectorInstrCostHelper(
         // important.
         UserToExtractIdx[U];
       }
+      if (UserToExtractIdx.empty())
```
Comment (sdesmalen-arm) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3279]: I don't see the link with the change in `IsExtractLaneEquivalentToZero`, is this related?

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 3]
```diff
@@ -3248,19 +3248,18 @@ InstructionCost AArch64TTIImpl::getVectorInstrCostHelper(
     // Check if the extractelement user is scalar fmul.
     auto IsUserFMulScalarTy = [](const Value *EEUser) {
       // Check if the user is scalar fmul.
-      const auto *BO = dyn_cast_if_present<BinaryOperator>(EEUser);
+      const auto *BO = dyn_cast<BinaryOperator>(EEUser);
```
Comment (sushgokh) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3251]: EEUser can be null and hence, dont change this

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 4]
```diff
@@ -3248,19 +3248,18 @@ InstructionCost AArch64TTIImpl::getVectorInstrCostHelper(
...
     // Check if the extract index is from lane 0 or lane equivalent to 0 for a
     // certain scalar type and a certain vector register width.
-    auto IsExtractLaneEquivalentToZero = [&](const unsigned &Idx,
-                                             const unsigned &EltSz) {
+    auto IsExtractLaneEquivalentToZero = [&](unsigned Idx, unsigned EltSz) {
       auto RegWidth =
           getRegisterBitWidth(TargetTransformInfo::RGK_FixedWidthVector)
               .getFixedValue();
-      return (Idx == 0 || (Idx * EltSz) % RegWidth == 0);
+      return RegWidth != 0 && (Idx == 0 || (Idx * EltSz) % RegWidth == 0);
```
Comment (sushgokh) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3262]: Could you please elaborate why the  `RegWidth` can be 0  in sme streaming functions? If there is some article which I can read about, you can point that.

Can you push the check inside?
`return Idx == 0 || (RegWidth != 0 && (Idx * EltSz) % RegWidth == 0)`

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp (hunk 1 above)
Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3258]: Yeah will do.

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp (hunk 3 above)
Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3251]: This function is called from users() in two cases. Can you explain why it can be null? Do you have a test case that shows it?

Comment (sushgokh) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3251]: sorry, my bad. You are right. (if_present was required in one of the earlier revisions of the original patch and that lasted till end)

**Review by sushgokh — APPROVED** [review]
LGTM

**Comment by davemgreen** [comment]
Thanks. The first part was split out into d106a39c3372b924668f203fedbce69aa986cf50.
